README.md
# cmtxgo

[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white&style=flat-square)](https://pkg.go.dev/github.com/dalikewara/cmtxgo)
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/dalikewara/cmtxgo)
![GitHub tag (latest SemVer)](https://img.shields.io/github/v/tag/dalikewara/cmtxgo)
![GitHub license](https://img.shields.io/github/license/dalikewara/cmtxgo)

**cmtxgo** generates cemtext format in Golang. Cemtext file format is a format used by banks to allow for batch
transactions. Banks adopted this format by default, but they may have different format style. This package allows you to
create a cemtext format by your own style—that means you can use it to meet your bank cemtext requirements.

> If you're working with JavaScript, you can use this module instead -> [https://github.com/dalikewara/vcmcemtex](https://github.com/dalikewara/vcmcemtex)

## Getting started

### Installation

You can use the `go install` method:

```bash
go install github.com/dalikewara/cmtxgo@latest
```

or, you can also use the `go get` method (DEPRECATED since `go1.17`):

```bash
go get github.com/dalikewara/cmtxgo
```

### Usage

A cemtext format commonly has 3 sections: `header`, `detail` and `footer`.

```text
(header) 0                 01BQL       MY NAME                   1111111004231633  230410
(detail) 1123-456157108231 530000001234S R SMITH                       TEST BATCH        062-000 12223123MY ACCOUNT      00001200
(detail) 1123-783 12312312 530000002200J K MATTHEWS                    TEST BATCH        062-000 12223123MY ACCOUNT      00000030
(detail) 1456-789   125123 530003123513P R JONES                       TEST BATCH        062-000 12223123MY ACCOUNT      00000000
(detail) 1121-232    11422 530000002300S MASLIN                        TEST BATCH        062-000 12223123MY ACCOUNT      00000000
(footer) 7999-999            000312924700031292470000000000                        000004
```

Every section detailed information about the transaction, and they can be separated by new line `\n` or just combined to
a single string. Both is valid, because it depends on the bank specification&mdash;some Indonesian bank uses single
string format, for example:

```text
0                 01BQL       MY NAME                   1111111004231633  2304101123-456157108231 530000001234S R SMITH                       TEST BATCH        062-000 12223123MY ACCOUNT      000012001123-783 12312312 530000002200J K MATTHEWS                    TEST BATCH        062-000 12223123MY ACCOUNT      000000301456-789   125123 530003123513P R JONES                       TEST BATCH        062-000 12223123MY ACCOUNT      000000001121-232    11422 530000002300S MASLIN                        TEST BATCH        062-000 12223123MY ACCOUNT      000000007999-999            000312924700031292470000000000                        000004
```

> Ref: [https://www.cemtexaba.com/aba-format](https://www.cemtexaba.com/aba-format)

This is the very basic usage of this package:

```go
cmtx := cmtxgo.NewCmtx()

// Set header
cmtx.SetHeader(&cmtxgo.Field{
    "myHeaderFieldName": &cmtxgo.Attribute{
        Order:  1,
        Type:   "lpz",
        Length: 11,
        Value:  "999",
    },
    "myHeaderFieldName2": &cmtxgo.Attribute{
        Order:  2,
        Type:   "rps",
        Length: 35,
        Value:  "My COMPANY",
    },
})

// Set detail
detailData := []map[string]interface{}{
    {
        "myDetailFieldName":  "123-456-789",
        "myDetailFieldName2": "John Doe",
        "myDetailFieldName3": "1000000",
    },
    {
        "myDetailFieldName":  "098-765-432",
        "myDetailFieldName2": "Adam Smith",
        "myDetailFieldName3": "1005500",
    },
    {
        "myDetailFieldName":  "123-098-456",
        "myDetailFieldName2": "Dali Kewara",
        "myDetailFieldName3": "204000",
    },
}
cmtx.SetDetail(&cmtxgo.Field{
    "myDetailFieldName": &cmtxgo.Attribute{
        Order:  1,
        Type:   "rps",
        Length: 15,
    },
    "myDetailFieldName2": &cmtxgo.Attribute{
        Order:  2,
        Type:   "rps",
        Length: 20,
    },
    "myDetailFieldName3": &cmtxgo.Attribute{
        Order:  3,
        Type:   "lpz",
        Length: 11,
    },
}, &detailData)

// Set footer
cmtx.SetFooter(&cmtxgo.Field{
    "myFooterFieldName": &cmtxgo.Attribute{
        Order:  1,
        Type:   "lpz",
        Length: 11,
        Value:  fmt.Sprintf("%v", len(detailData)),
    },
    "myFooterFieldName2": &cmtxgo.Attribute{
        Order:  2,
        Type:   "lpz",
        Length: 35,
        Value:  "2209500",
    },
})

// Generate the cemtext format as string
cmtxStr := cmtx.Generate()
```

The output is:

```text
00000000999My COMPANY                         123-456-789    John Doe            00001000000098-765-432    Adam Smith          00001005500123-098-456    Dali Kewara         000002040000000000000300000000000000000000000000002209500
```

If you want to save the cemtext format into a file, just use `GenerateToFile` method instead:

```go
err := cmtx.GenerateToFile("/path/to/filename.ctx")
if err != nil {
    panic(err)
}
```

### Attribute

- `Order`: sorts the generation of the fields. Example:

```go
cmtx.SetHeader(&cmtxgo.Field{
    "first": &cmtxgo.Attribute{
        Order:  2,
        Type:   "lps",
        Length: 6,
        Value:  "Second",
    },
    "second": &cmtxgo.Attribute{
        Order:  1,
        Type:   "lps",
        Length: 5,
        Value:  "First",
    },
})
// Output: "FirstSecond"
```

- `Type`: formats the field value based on padding type. Available padding type: `lps`, `lpz`, `rps`, `rpz`. 
Default is `rps`.
  - `lps`: Left Padding Space (ex: "  john doe")
  - `lpz`: Left Padding Zero (ex: "0000000123")
  - `rps`: Right Padding Space (ex: "john doe   ")
  - `rpz`: Right Padding Zero (ex: "1230000000")
- `Length`: sets max length to the field value.
- `Value`: sets field value. This only effect on `header` & `footer` section.
For `detail` section, the field value got from the matched field name in detail data.
If Value is empty, DefaultValue will be used.
- `DefaultValue`: sets the default value to the field if the Value is empty. Default is "".
- `RemoveAllChars`: removes specified characters from the field value. Example:

```go
cmtx.SetFooter(&cmtxgo.Field{
    "total_amount": &cmtxgo.Attribute{
        Order:          1,
        Type:           "lpz",
        Length:         10,
        Value:          "1.000.000.000",
        RemoveAllChars: ".",
    },
})
// Output: "1000000000"
```

- `ReplaceAllChars`: ReplaceAllChars replaces specified characters with the new one from the field value. Example:

```go
cmtx.SetFooter(&cmtxgo.Field{
    "total_amount": &cmtxgo.Attribute{
        Order:           1,
        Type:            "lpz",
        Length:          13,
        Value:           "1.000.000.000",
        ReplaceAllChars: [2]string{".", ","},
    },
})
// Output: "1,000,000,000"
```

### Option

- `MaxLengthPerSection`: sets max length on every section value.
- `AddCharPerSection`: adds specified character at the end of the value on every section. Keep in mind that this will add 
new character, so it will break the value length setting.

## Release

### Changelog

Read at [CHANGELOG.md](https://github.com/dalikewara/cmtxgo/blob/master/CHANGELOG.md)

### Credits

Copyright &copy; 2022 [Dali Kewara](https://www.dalikewara.com)

### License

[MIT License](https://github.com/dalikewara/cmtxgo/blob/master/LICENSE)


CHANGELOG.md
# Changelogs

### 2022

- v1.0.2 (2022-09-13)
    - Update readme

- v1.0.1 (2022-09-12)
    - Change license to MIT License 

- v1.0.0 (2022-02-17)
    - Initial release


generate_test.go
package cmtxgo_test

import (
	"fmt"
	"github.com/dalikewara/cmtxgo"
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestAttribute_Order(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetHeader(&cmtxgo.Field{
		"first": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 5,
			Value:  "First",
		},
		"third": &cmtxgo.Attribute{
			Order:  3,
			Type:   "lps",
			Length: 5,
			Value:  "Third",
		},
		"second": &cmtxgo.Attribute{
			Order:  2,
			Type:   "lps",
			Length: 6,
			Value:  "Second",
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "FirstSecondThird", cmtxStr)
}

func TestAttribute_Type(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	t.Run("Left Padding", func(t *testing.T) {
		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 10,
				Value:  "Company",
			},
		})
		cmtxStr := cmtx.Generate()
		assert.Equal(t, "   Company", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lpz",
				Length: 10,
				Value:  "1",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "0000000001", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 10,
				Value:  "Company",
			},
			"record": &cmtxgo.Attribute{
				Order:  2,
				Type:   "lpz",
				Length: 10,
				Value:  "1",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "   Company0000000001", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 4,
				Value:  "Company",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "Comp", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lpz",
				Length: 4,
				Value:  "12345",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "1234", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 7,
				Value:  "Company",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "Company", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lpz",
				Length: 5,
				Value:  "12345",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "12345", cmtxStr)
	})

	t.Run("Right Padding", func(t *testing.T) {
		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rps",
				Length: 10,
				Value:  "Company",
			},
		})
		cmtxStr := cmtx.Generate()
		assert.Equal(t, "Company   ", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rpz",
				Length: 10,
				Value:  "1",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "1000000000", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rps",
				Length: 10,
				Value:  "Company",
			},
			"record": &cmtxgo.Attribute{
				Order:  2,
				Type:   "rpz",
				Length: 10,
				Value:  "1",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "Company   1000000000", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rps",
				Length: 4,
				Value:  "Company",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "Comp", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rpz",
				Length: 4,
				Value:  "12345",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "1234", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rps",
				Length: 7,
				Value:  "Company",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "Company", cmtxStr)

		cmtx.SetHeader(&cmtxgo.Field{
			"record": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rpz",
				Length: 5,
				Value:  "12345",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "12345", cmtxStr)
	})
}

func TestAttribute_Length(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 4,
			Value:  "Company",
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, 4, len(cmtxStr))
	assert.Equal(t, "Comp", cmtxStr)

	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lpz",
			Length: 6,
			Value:  "1234567",
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, 6, len(cmtxStr))
	assert.Equal(t, "123456", cmtxStr)

	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 7,
			Value:  "Company",
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, 7, len(cmtxStr))
	assert.Equal(t, "Company", cmtxStr)

	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lpz",
			Length: 7,
			Value:  "1234567",
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, 7, len(cmtxStr))
	assert.Equal(t, "1234567", cmtxStr)
}

func TestAttribute_Value(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 7,
			Value:  "Company",
		},
		"record": &cmtxgo.Attribute{
			Order:  2,
			Type:   "lpz",
			Length: 10,
			Value:  "1",
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "Company0000000001", cmtxStr)
}

func TestAttribute_DefaultValue(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetHeader(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:        1,
			Type:         "lps",
			Length:       7,
			DefaultValue: "Company",
		},
		"record": &cmtxgo.Attribute{
			Order:        2,
			Type:         "lpz",
			Length:       10,
			Value:        "2",
			DefaultValue: "1",
		},
		"branch": &cmtxgo.Attribute{
			Order:        3,
			Type:         "lpz",
			Length:       10,
			Value:        "",
			DefaultValue: "3",
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "Company00000000020000000003", cmtxStr)
}

func TestAttribute_RemoveAllChars(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetFooter(&cmtxgo.Field{
		"total_amount": &cmtxgo.Attribute{
			Order:          1,
			Type:           "lpz",
			Length:         11,
			Value:          "1.000.000.000",
			RemoveAllChars: ".",
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "01000000000", cmtxStr)

	cmtx.SetFooter(&cmtxgo.Field{
		"total_amount": &cmtxgo.Attribute{
			Order:          1,
			Type:           "lpz",
			Length:         12,
			Value:          "1000000,000",
			RemoveAllChars: ".",
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, "01000000,000", cmtxStr)
}

func TestAttribute_ReplaceAllChars(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetFooter(&cmtxgo.Field{
		"total_amount": &cmtxgo.Attribute{
			Order:           1,
			Type:            "lpz",
			Length:          11,
			Value:           "1.000.000.000",
			ReplaceAllChars: [2]string{".", ","},
		},
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "1,000,000,0", cmtxStr)

	cmtx.SetFooter(&cmtxgo.Field{
		"total_amount": &cmtxgo.Attribute{
			Order:           1,
			Type:            "lpz",
			Length:          12,
			Value:           "1000000,000",
			ReplaceAllChars: [2]string{",", "."},
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, "01000000.000", cmtxStr)

	cmtx.SetFooter(&cmtxgo.Field{
		"total_amount": &cmtxgo.Attribute{
			Order:           1,
			Type:            "lpz",
			Length:          12,
			Value:           "1000000000",
			ReplaceAllChars: [2]string{",", "."},
		},
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, "001000000000", cmtxStr)
}

func TestOption_MaxLengthPerSection(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetFooter(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 10,
			Value:  "Company",
		},
	})
	cmtx.SetOption(&cmtxgo.Option{
		MaxLengthPerSection: 20,
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, 20, len(cmtxStr))
	assert.Equal(t, "   Company          ", cmtxStr)

	cmtx.SetFooter(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 20,
			Value:  "Company",
		},
	})
	cmtx.SetOption(&cmtxgo.Option{
		MaxLengthPerSection: 20,
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, 20, len(cmtxStr))
	assert.Equal(t, "             Company", cmtxStr)
}

func TestOption_AddCharPerSection(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	cmtx.SetFooter(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 10,
			Value:  "Company",
		},
	})
	cmtx.SetOption(&cmtxgo.Option{
		AddCharPerSection: "\n",
	})
	cmtxStr := cmtx.Generate()
	assert.Equal(t, "   Company\n", cmtxStr)

	cmtx.SetFooter(&cmtxgo.Field{
		"name": &cmtxgo.Attribute{
			Order:  1,
			Type:   "lps",
			Length: 10,
			Value:  "Company",
		},
	})
	cmtx.SetOption(&cmtxgo.Option{
		AddCharPerSection: "ADDED",
	})
	cmtxStr = cmtx.Generate()
	assert.Equal(t, "   CompanyADDED", cmtxStr)
}

func TestDetail(t *testing.T) {
	cmtx := cmtxgo.NewCmtx()
	t.Run("Plain", func(t *testing.T) {
		cmtx.SetDetail(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 10,
			},
			"username": &cmtxgo.Attribute{
				Order:  2,
				Type:   "lps",
				Length: 10,
			},
		}, &[]map[string]interface{}{
			{
				"name":     "John Doe",
				"username": "john_doe",
			},
			{
				"name":     "Adam Smith",
				"username": "adam_smith",
			},
		})
		cmtxStr := cmtx.Generate()
		assert.Equal(t, "  John Doe  john_doeAdam Smithadam_smith", cmtxStr)

		cmtx.SetDetail(&cmtxgo.Field{
			"name": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lps",
				Length: 10,
			},
			"username": &cmtxgo.Attribute{
				Order:  2,
				Type:   "lps",
				Length: 10,
			},
		}, &[]map[string]interface{}{
			{
				"name": "John Doe",
			},
			{
				"name":     "Dali Kewara",
				"username": "dalikewara",
			},
			{
				"name": "Adam Smith",
			},
		})
		cmtxStr = cmtx.Generate()
		assert.Equal(t, "  John Doe          Dali KewardalikewaraAdam Smith          ", cmtxStr)
	})
}

func TestExample(t *testing.T) {
	t.Run("Basic Usage", func(t *testing.T) {
		cmtx := cmtxgo.NewCmtx()
		cmtx.SetHeader(&cmtxgo.Field{
			"myHeaderFieldName": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lpz",
				Length: 11,
				Value:  "999",
			},
			"myHeaderFieldName2": &cmtxgo.Attribute{
				Order:  2,
				Type:   "rps",
				Length: 35,
				Value:  "My COMPANY",
			},
		})
		detailData := []map[string]interface{}{
			{
				"myDetailFieldName":  "123-456-789",
				"myDetailFieldName2": "John Doe",
				"myDetailFieldName3": "1000000",
			},
			{
				"myDetailFieldName":  "098-765-432",
				"myDetailFieldName2": "Adam Smith",
				"myDetailFieldName3": "1005500",
			},
			{
				"myDetailFieldName":  "123-098-456",
				"myDetailFieldName2": "Dali Kewara",
				"myDetailFieldName3": "204000",
			},
		}
		cmtx.SetDetail(&cmtxgo.Field{
			"myDetailFieldName": &cmtxgo.Attribute{
				Order:  1,
				Type:   "rps",
				Length: 15,
			},
			"myDetailFieldName2": &cmtxgo.Attribute{
				Order:  2,
				Type:   "rps",
				Length: 20,
			},
			"myDetailFieldName3": &cmtxgo.Attribute{
				Order:  3,
				Type:   "lpz",
				Length: 11,
			},
		}, &detailData)
		cmtx.SetFooter(&cmtxgo.Field{
			"myFooterFieldName": &cmtxgo.Attribute{
				Order:  1,
				Type:   "lpz",
				Length: 11,
				Value:  fmt.Sprintf("%v", len(detailData)),
			},
			"myFooterFieldName2": &cmtxgo.Attribute{
				Order:  2,
				Type:   "lpz",
				Length: 35,
				Value:  "2209500",
			},
		})
		cmtxStr := cmtx.Generate()
		err := cmtx.GenerateToFile("tmp/cemtext.ctx")
		assert.Nil(t, err)
		assert.Equal(t, "00000000999My COMPANY                         123-456-789    John Doe            00001000000098-765-432    Adam Smith          00001005500123-098-456    Dali Kewara         000002040000000000000300000000000000000000000000002209500", cmtxStr)
	})
}


generate.go
package cmtxgo

import (
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
)

type Attribute struct {
	// Order sorts the generation of the fields. Example:
	//
	//cmtx.SetHeader(&cmtxgo.Field{
	//	"first": &cmtxgo.Attribute{
	//		Order:  2,
	//		Type:   "lps",
	//		Length: 6,
	//		Value:  "Second",
	//	},
	//	"second": &cmtxgo.Attribute{
	//		Order:  1,
	//		Type:   "lps",
	//		Length: 5,
	//		Value:  "First",
	//	},
	//})
	//
	// Output: "FirstSecond"
	Order int `json:"order"`

	// Type formats the field value based on padding type. Available padding type: `lps`, `lpz`, `rps`, `rpz`.
	// Default is `rps`.
	//
	// `lps`: Left Padding Space (ex: "  john doe")
	//
	// `lpz`: Left Padding Zero (ex: "0000000123")
	//
	// `rps`: Right Padding Space (ex: "john doe   ")
	//
	// `rpz`: Right Padding Zero (ex: "1230000000")
	Type string `json:"type"`

	// Length sets max length to the field value.
	Length int `json:"length"`

	// Value sets field value. This only effect on `header` & `footer` section.
	// For `detail` section, the field value got from the matched field name in detail data.
	// If Value is empty, DefaultValue will be used.
	Value string `json:"value"`

	// DefaultValue sets the default value to the field if the Value is empty. Default is "".
	DefaultValue string `json:"default_value"`

	// RemoveAllChars removes specified characters from the field value. Example:
	//
	//cmtx.SetFooter(&cmtxgo.Field{
	//	"total_amount": &cmtxgo.Attribute{
	//		Order:          1,
	//		Type:           "lpz",
	//		Length:         10,
	//		Value:          "1.000.000.000",
	//		RemoveAllChars: ".",
	//	},
	//})
	//
	// Output: "1000000000"
	RemoveAllChars string `json:"remove_all_chars"`

	// ReplaceAllChars replaces specified characters with the new one from the field value.
	// Example:
	//
	//cmtx.SetFooter(&cmtxgo.Field{
	//	"total_amount": &cmtxgo.Attribute{
	//		Order:           1,
	//		Type:            "lpz",
	//		Length:          13,
	//		Value:           "1.000.000.000",
	//		ReplaceAllChars: [2]string{".", ","},
	//	},
	//})
	//
	// Output: "1,000,000,000"
	ReplaceAllChars [2]string `json:"replace_all_chars"`
}

type Option struct {
	// MaxLengthPerSection sets max length on every section value.
	MaxLengthPerSection int `json:"max_length_per_section"`

	// AddCharPerSection adds specified character at the end of the value on every section.
	// Keep in mind that this will add new character, so it will break
	// the value length setting.
	AddCharPerSection string `json:"add_char_per_section"`
}

type Field map[string]*Attribute

type cmtx struct {
	header     *Field
	footer     *Field
	detail     *Field
	detailData *[]map[string]interface{}
	option     *Option
}

var orderSep = " __{{[[((0RD3R_53P))]]}}__ "

// NewCmtx generates new cmtx object.
func NewCmtx() *cmtx {
	return &cmtx{}
}

// SetHeader sets the cemtext header.
func (c *cmtx) SetHeader(field *Field) {
	c.header = field
}

// SetDetail sets the cemtext detail.
func (c *cmtx) SetDetail(field *Field, detailData *[]map[string]interface{}) {
	c.detail = field
	c.detailData = detailData
}

// SetFooter sets the cemtext footer.
func (c *cmtx) SetFooter(field *Field) {
	c.footer = field
}

// SetOption sets the generation option of each section.
func (c *cmtx) SetOption(option *Option) {
	c.option = option
}

// Generate generates cemtext format as string.
func (c *cmtx) Generate() string {
	header := ""
	detail := ""
	footer := ""

	var wg sync.WaitGroup

	wg.Add(3)

	go func() {
		defer wg.Done()
		header = generateHeader(c.header, c.option)
	}()
	go func() {
		defer wg.Done()
		detail = generateDetail(c.detail, c.detailData, c.option)
	}()
	go func() {
		defer wg.Done()
		footer = generateFooter(c.footer, c.option)
	}()

	wg.Wait()

	return fmt.Sprintf("%s%s%s", header, detail, footer)
}

// GenerateToFile generates cemtext format and save it into a file.
func (c *cmtx) GenerateToFile(output string) error {
	cmtxStr := c.Generate()
	file, err := os.Create(output)
	if err != nil {
		return err
	} else {
		_, errWriteString := file.WriteString(cmtxStr)
		if errWriteString != nil {
			return errWriteString
		}
	}
	err = file.Close()
	if err != nil {
		return err
	}
	return nil
}

func generateCommon(field *Field) string {
	if field == nil {
		return ""
	}
	var orders []string
	for key, attr := range *field {
		orders = append(orders, fmt.Sprintf("%v%s%s%s", attr.Order, key, orderSep, generateValue(attr)))
	}
	sort.Strings(orders)
	str := ""
	for _, s := range orders {
		str = str + strings.Split(s, orderSep)[1]
	}
	return str
}

func generateHeader(field *Field, option *Option) string {
	str := generateCommon(field)
	if str != "" && option != nil {
		str = formatMaxLength(str, option.MaxLengthPerSection)
		if option.AddCharPerSection != "" {
			str = addCharAtTrail(str, option.AddCharPerSection)
		}
	}
	return str
}

func generateDetail(field *Field, detailData *[]map[string]interface{}, option *Option) string {
	if field == nil || detailData == nil {
		return ""
	}
	str := ""
	for _, v := range *detailData {
		dataField := interface{}(*field).(Field)
		for n, _ := range dataField {
			dataField[n].Value = ""
			if v[n] != "" && v[n] != nil {
				dataField[n].Value = fmt.Sprintf("%v", v[n])
			}
		}
		cStr := generateCommon(field)
		if cStr != "" && option != nil {
			cStr = formatMaxLength(cStr, option.MaxLengthPerSection)
			if option.AddCharPerSection != "" {
				cStr = addCharAtTrail(cStr, option.AddCharPerSection)
			}
		}
		str = str + cStr
	}
	return str
}

func generateFooter(field *Field, option *Option) string {
	return generateHeader(field, option)
}

func generateValue(attr *Attribute) string {
	val := attr.DefaultValue
	if attr.Value != "" {
		val = attr.Value
	}
	if attr.RemoveAllChars != "" {
		val = removeAllChars(val, attr.RemoveAllChars)
	}
	if attr.ReplaceAllChars[0] != "" && attr.ReplaceAllChars[1] != "" {
		val = replaceAllChars(val, attr.ReplaceAllChars[0], attr.ReplaceAllChars[1])
	}
	return formatValue(val, attr)
}

func formatMaxLength(val string, maxLength int) string {
	if maxLength != 0 {
		return rightPad(val, " ", maxLength)
	}
	return val
}

func formatValue(val string, attr *Attribute) string {
	switch attr.Type {
	case "lps":
		return leftPad(val, " ", attr.Length)
	case "lpz":
		return leftPad(val, "0", attr.Length)
	case "rpz":
		return rightPad(val, "0", attr.Length)
	default:
		return rightPad(val, " ", attr.Length)
	}
}

func leftPad(val string, pad string, length int) string {
	for i := len(val); i < length; i++ {
		val = pad + val
	}
	return val[:length]
}

func rightPad(val string, pad string, length int) string {
	for i := len(val); i < length; i++ {
		val = val + pad
	}
	return val[:length]
}

func addCharAtTrail(val string, char string) string {
	return fmt.Sprintf("%s%s", val, char)
}

func replaceAllChars(val, old, new string) string {
	return strings.ReplaceAll(val, old, new)
}

func removeAllChars(val, char string) string {
	return strings.ReplaceAll(val, char, "")
}
